articles/node_lifecycle.md

```diff
@@ -0,0 +1,176 @@
+
+[To discuss: Should there be a callback for when configuration is changed, e.g. altered parameters, new topics, etc.?]
+
+While in this state, the node will not receive any execution time to read topics, perform processing of data, respond to functional service requests, etc.
+
+In the inactive state, any data that arrives on topics will not be read. Whether or not data is queued will depend on the QoS parameters of the topics involved. [In other words, to queue data or not is up to each topic.]
+
+Any service requests to a node in the inactive state will not be answered (to the caller, they will fail immediately).
+
+Entering the `inactive` state from the `created` state requires running the node's `onInitialised` method. The node uses this to set up any resources it must hold throughout its life (irrespective of if it is active or inactive). As examples, such resources may include topic publications and subscriptions, memory that is held continuously, and initialising configuration parameters.
+
+Entering the `inactive` state from the `active` state requires running the node's `onDeactivated` method. See the `active` state description for more information.
+
+Entering the `inactive` state from the `fatal error` state requires running the node's `onReset` method. See the `fatal error` state description for more information.
+
+If an error occurs while in this state that cannot be handled by the node, then the transition to the `fatal error` state is taken.
+
+The node may also exit this state by being destroyed, in which case the `destroy` transition is taken.
+
+Implementation note: Causing the transition to the destroyed state can often be mapped to deletion of the node instance.
+
+### Active state
+
+This is the main state of the node's life cycle. While in this state, the node performs any processing, responds to service requests, reads and processes data, produces output, etc.
+
+Entering the `active` state requires running the node's `onActivated` method. This is used to prepare the node for execution. This may include acquiring resources that are only held while the node is actually active, such as access to hardware. Ideally, no preparation that requires significant time (such as lengthy hardware initialisation) should be performed in this callback.
+
+To shut down a node, it must first be deactivated. This will cause the `onDeactivate` method to be called, allowing the node to clean up any resources it only holds while active.
+
+If an error that cannot be handled by the node/system occurs in this state, the transition to the `fatal error` state is taken.
+
+### Fatal Error state
+
+Known errors that can occur in a node should be considered designed behaviour of the system and so should be handled by the node where the error occurred and, if necessary, the surrounding nodes. The reason for this is that the system knows best how to handle errors that occur in its behaviour.
```

Comment:
> I think it may be more complicated than this part of the document lets on.

Yes it is. My main focus is on such executors and I definitely *do not* want nodes to do *any* form of error recovery. An "error" is by definition something the node should not recover for. Otherwise, it's a nominal situation. The drivers are typical: if they're not there, then the node should yell. An active driver should produce data, period. Most systems will be non-functional and assuming that you have a node-specific notification channel for the error cases IMO negates a big part of the benefit of a livecycle state machine.

[Rock](https://rock-robotics.org) did not standardize at all on ROS1 in part because of this completely "nodes are meant to be magical" design decision (the other one is the inability to dynamically change the connection graph). I just hope that ROS2 design documents won't try to push people towards the same behaviour.